### PR TITLE
GH-90699: disallow `_Py_IDENTIFIER` in core code 

### PR DESCRIPTION
**Pull request by kumaraditya303**
* Issue: gh-90699

Automerge-Triggered-By: GH:ericsnowcurrently

### CODE REVIEW

File: Programs/_testembed.c [hunk 1]
```diff
@@ -1891,7 +1890,11 @@ static int test_unicode_id_init(void)
 {
     // bpo-42882: Test that _PyUnicode_FromId() works
     // when Python is initialized multiples times.
-    _Py_IDENTIFIER(test_unicode_id_init);
+
+    static _Py_Identifier PyId_test_unicode_id_init = {
+        .string = "test_unicode_id_init",
+        .index = -1,
+    };
```
Comment (ericsnowcurrently) [inline Programs/_testembed.c:1900]: I suppose using `_Py_IDENTIFIER()` here was an implicit test of sorts, especially considering we've removed all other uses of it.  However, manually expanding the macro here is probably fine since we're actually testing `_PyUnicode_FromId()` and we are looking to possibly remove `_Py_IDENTIFIER()` from the C-API completely.

On the other hand, we no longer use `_Py_IDENTIFIER()` anywhere else in the repo, and the value of continuing to test `_PyUnicode_FromId()` across multiple runtime initializations is mostly gone, so maybe we should just drop `test_unicode_id_init()`.

On the other other hand, we aren't getting rid of `_Py_IDENTIFIER()` yet (due to external usage [1][2]).  As long as we have it we should be testing it.  That should probably be in test_capi.py (and Modules/_testinternalcapi.c) though, rather than in test_embed.py (and Programs/_testembed.c).  If we do keep it then `test_unicode_id_init()` would probably be worth keeping too.

Bottom line: we should probably determine what we're going to do with `_Py_IDENTIFIER()` before we merge this change.


[1] https://github.com/python/cpython/issues/90699#issuecomment-1093943448
[2] https://mail.python.org/archives/list/python-dev@python.org/message/7RMLIJHUWVBZFV747TFEHOE6LNBVQSMM/

Comment (kumaraditya303) [reply Programs/_testembed.c:1900]: I went with expanding the macro as it seemed the least invasive change. `_Py_IDENTIFIER` will be sort of frozen in CPython now so expanding macro is IMO okay. `_Py_IDENTIFIER` should be deprecated as it is not compatible with sub-interpreters and per interpreter GIL and all that but unfortunately but I don't have the time to do that. All I want to do in this PR is to disable using `_Py_IDENTIFIER` internally in the code base.

Comment (kumaraditya303) [reply Programs/_testembed.c:1900]: Let's keep the test for now and just expand the macro so it is will still be  tested. This test will be removed in PR which **actually** removes `_Py_IDENTIFIER` together.

Comment (ericsnowcurrently) [reply Programs/_testembed.c:1900]: I'm fine with leaving the test for now.  Please add a comment indicating why we are manually expanding `_Py_IDENTIFIER()` there.  Otherwise LGTM.

Comment (kumaraditya303) [reply Programs/_testembed.c:1900]: done

**Review by ericsnowcurrently — APPROVED** [review]
LGTM

**Comment by ericsnowcurrently** [comment]
Thanks for working on this, @kumaraditya303!
